src/lib/HST/Frontend/Syntax.hs

```diff
@@ -0,0 +1,120 @@
+module HST.Frontend.Syntax where
+
+data Module s l t = Module [Decl s l t]
+  deriving (Eq, Show)
+
+data Decl s l t = DataDecl (DeclHead s) [ConDecl s t]
+                | TypeSig (SrcSpan s) [Name s] t
+                | FunBind (SrcSpan s) [Match s l t]
+                | PatBind (SrcSpan s) (Pat s l) (Rhs s l t) (Maybe (Binds s l t))
+  deriving (Eq, Show)
+
+data DeclHead s = DHead (Name s)
+                | DHInfix (Name s)
+                | DHParen (DeclHead s)
+                | DHApp (DeclHead s)
+  deriving (Eq, Show)
+
+data ConDecl s t = ConDecl (Name s) [t]
+                 | InfixConDecl t (Name s) t
+                 | RecDecl (Name s)
+  deriving (Eq, Show)
+
+data Binds s l t = BDecls (SrcSpan s) [Decl s l t]
+  deriving (Eq, Show)
+
+data Match s l t = Match (SrcSpan s) (Name s) [Pat s l] (Rhs s l t) (Maybe (Binds s l t))
+                 | InfixMatch (SrcSpan s) (Pat s l) (Name s) [Pat s l] (Rhs s l t) (Maybe (Binds s l t))
+  deriving (Eq, Show)
+
+data Rhs s l t = UnGuardedRhs (SrcSpan s) (Exp s l t)
+               | GuardedRhss (SrcSpan s) [GuardedRhs s l t]
+  deriving (Eq, Show)
+
+data GuardedRhs s l t = GuardedRhs (SrcSpan s) [Stmt s l t] (Exp s l t)
+  deriving (Eq, Show)
+
+data Boxed = Boxed
+           | Unboxed
+  deriving (Eq, Show)
+
+data Exp s l t = Var (SrcSpan s) (QName s)
```

Comment:
We could use type families to reduce the number of type arguments. This would improve the readability of code that uses the AST.

```haskell
{-# LANGUAGE TypeFamilies #-}

type family SrcSpan a
type family Literal a
type family TypeExp a

data Exp a = … | Lit (SrcSpan a) (Literal a) | …
```

Type family instances for HSE could look as follows, for example.

```haskell
data HSE
type instance SrcSpan HSE = HSE.SrcSpanInfo
type instance Literal HSE = HSE.Literal HSE.SrcSpanInfo
type instance TypeExp HSE = HSE.Type HSE.SrcSpanInfo
```

Instead of `Exp HSE.SrcSpanInfo (HSE.Literal HSE.SrcSpanInfo) (HSE.Type HSE.SrcSpanInfo)` we would now have to write `Exp HSE` only. However, I don't have enough experience with type families to tell  whether there would be any drawbacks. @ichistmeinname do you think that a type family based approach would work in this case?

Reply:
Not sure I have the best overview yet, but I couldn't find concrete instantiations of `l` and `t`, and, thus, it's hard to argue what to do here. We should discuss that later.

Concerning the type family: I have seen a a lot of compiler/AST-based implementations, but never saw the type family approach. So I wouldn't opt for such a solution.

Reply:
I have implemented these type families in 3073104. The biggest advantage is the reduction of type signature sizes, but on the other hand, deriving `Eq` and `Show` instances seems to be more complicated. Most importantly, it doesn't seem to be possible to derive `Eq` and `Show` instances directly in `Syntax.hs` for any type directly or indirectly containing one of these type families (excluding `Eq` instances for types only containing the `SrcSpanType` family, which work because of the custom equality). We can instead derive these instances using the `StandaloneDeriving` pragma with concrete types in `FromHSE.hs`, but we have to do that for almost every single type in `Syntax.hs`.
The most important implication of this is that when we need an `Eq` or `Show` instance for one specific Type from `Syntax.hs` (for example `Exp`), we have to add that specific instance as a constraint (for example `Eq (S.Exp a) =>`), which requires the `FlexibleContexts` pragma. We currently only use `Eq` for `Exp`, but this might become a problem if we want to compare other constructs or use `Show` or other classes, because those constraints would accumulate fast. For example, if we wanted to use `Eq` and `Show` instances for `Decl`, `Alt` and `Exp`, we would need a `(Eq (S.Decl a), Eq (S.Alt a), Eq (S.Exp a), Show (S.Decl a), Show (S.Alt a), Show (S.Exp a))` constraint. Do you have an idea on how to solve this problem?

Reply:
Maybe a type classes "wrapping" is necessary?
https://stackoverflow.com/questions/25219162/type-class-constraint-on-type-family-instances
https://downloads.haskell.org/~ghc/7.0.1/docs/html/users_guide/type-families.html (Sec. 7.7.2.1.1.)

Reply:
After some tries, I was able to implement something similar to this in d088904, which seems to work well.